File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/Visitors.java
package ro.pub.cs.diploma;

import com.intellij.psi.*;

import java.util.ArrayList;
import java.util.List;

class Visitors {
  static List<PsiReturnStatement> extractReturnStatements(PsiCodeBlock block) {
    final List<PsiReturnStatement> returnStatements = new ArrayList<>();
    block.accept(new JavaRecursiveElementWalkingVisitor() {
      @Override
      public void visitReturnStatement(PsiReturnStatement statement) {
        super.visitReturnStatement(statement);
        returnStatements.add(statement);
      }

      @Override
      public void visitClass(PsiClass aClass) {
      }

      @Override
      public void visitLambdaExpression(PsiLambdaExpression expression) {
      }
    });
    return returnStatements;
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/Visitors.kt
package ro.pub.cs.diploma

import com.intellij.psi.*

fun PsiCodeBlock.extractReturnStatements(): List<PsiReturnStatement> {
  val returnStatements = ArrayList<PsiReturnStatement>()
  accept(object : JavaRecursiveElementWalkingVisitor() {
    override fun visitReturnStatement(statement: PsiReturnStatement) {
      super.visitReturnStatement(statement)
      returnStatements.add(statement)
    }

    override fun visitClass(aClass: PsiClass) {}

    override fun visitLambdaExpression(expression: PsiLambdaExpression) {}
  })
  return returnStatements
}
